Repository/ContainerRepositoryFactory.php

```diff
@@ -0,0 +1,78 @@
+{
+    /** @var ObjectRepository[] */
+    private $managedRepositories = [];
+
+    /** @var ContainerInterface|null */
+    private $container;
+
+    /**
+     * @param ContainerInterface $container A service locator containing the repositories
+     */
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepository(DocumentManager $documentManager, $documentName)
+    {
+        $metadata             = $documentManager->getClassMetadata($documentName);
+        $customRepositoryName = $metadata->customRepositoryClassName;
+
+        if ($customRepositoryName !== null) {
+            // fetch from the container
+            if ($this->container && $this->container->has($customRepositoryName)) {
+                $repository = $this->container->get($customRepositoryName);
+
+                if (! $repository instanceof DocumentRepository) {
+                    throw new \RuntimeException(sprintf('The service "%s" must extend DocumentRepository (or a base class, like ServiceDocumentRepository).', $customRepositoryName));
+                }
+
+                return $repository;
+            }
+
+            // if not in the container but the class/id implements the interface, throw an error
+            if (is_a($customRepositoryName, ServiceDocumentRepositoryInterface::class, true)) {
+                throw new \RuntimeException(sprintf('The "%s" document repository implements "%s", but its service could not be found. Make sure the service exists and is tagged with "%s".', $customRepositoryName, ServiceDocumentRepositoryInterface::class, ServiceRepositoryCompilerPass::REPOSITORY_SERVICE_TAG));
+            }
+
+            if (! class_exists($customRepositoryName)) {
+                throw new \RuntimeException(sprintf('The "%s" document has a repositoryClass set to "%s", but this is not a valid class. Check your class naming. If this is meant to be a service id, make sure this service exists and is tagged with "%s".', $metadata->name, $customRepositoryName, ServiceRepositoryCompilerPass::REPOSITORY_SERVICE_TAG));
+            }
+
+            // allow the repository to be created below
+        }
+
+        return $this->getOrCreateRepository($documentManager, $metadata);
+    }
+
+    private function getOrCreateRepository(DocumentManager $documentManager, ClassMetadata $metadata)
+    {
+        $repositoryHash = $metadata->getName() . spl_object_hash($documentManager);
+        if (isset($this->managedRepositories[$repositoryHash])) {
+            return $this->managedRepositories[$repositoryHash];
+        }
+
+        $repositoryClassName = $metadata->customRepositoryClassName ?: $documentManager->getConfiguration()->getDefaultRepositoryClassName();
+
+        return $this->managedRepositories[$repositoryHash] = new $repositoryClassName($documentManager, $documentManager->getUnitOfWork(), $metadata);
```

Comment:
Shouldn't this defer repository creation to `DocumentManager` (which would then defer creation to repository factory)? Right now custom repositories with additional dependencies will be broken if I'm not mistaken

Reply:
I didn't want to wrap the default factory, but could certainly do so. Either way, custom repositories with additional dependencies **are** broken, which is why this PR will allow you to declare them as services to be fetched by this new factory.

Reply:
Ahhh all right, I think I've wrapped my head around this. This class could extend [`AbstractRepositoryFactory`](https://github.com/doctrine/mongodb-odm/blob/1.3.x/lib/Doctrine/ODM/MongoDB/Repository/AbstractRepositoryFactory.php), this way you'll get rid of managing instances here. WDYT?

Reply:
I'm not sure we want this. The original implementation always fetches the repository service from the container without it caching it locally. Only repositories that have to be instantiated are cached. That's the reason why I didn't use the abstract factory from the start since I was trying to keep this behaviour intact. It could make a difference when a service isn't shared, e.g. a new instance is returned when getting it from the container. AFAIK that also holds true when using `ServiceLocatorTagPass` to register a service locator with a bunch of services.